src/crimson/os/futurized_store.h

```diff
@@ -185,4 +186,251 @@ inline void intrusive_ptr_release(FuturizedStore::OmapIterator* iter) {
+  }
+
+  /**
+   * _OmapIterator
+   *
+   * Proxies OmapIterator operations to store's core.  Assumes that
+   * syncronous methods are safe to call directly from calling core
+   * since remote store should only be touching that memory during
+   * a method invocation.
+   *
+   * TODO: We don't really need OmapIterator at all, replace it with
+   * an appropriately paged omap_get_values variant.
+   */
+  class _OmapIterator : public OmapIterator {
+    using fref_t = seastar::foreign_ptr<OmapIteratorRef>;
+    const core_id_t core;
+    fref_t impl;
+
+    template <typename Method, typename... Args>
+    auto proxy(Method method, Args&&... args) {
+      return proxy_method_on_core(
+	core, *impl, method, std::forward<Args>(args)...);
+    }
+
+  public:
+    _OmapIterator(core_id_t core, fref_t &&impl)
+      : core(core), impl(std::move(impl)) {}
+
+    seastar::future<> seek_to_first() final {
+      return proxy(&OmapIterator::seek_to_first);
+    }
+    seastar::future<> upper_bound(const std::string &after) final {
+      return proxy(&OmapIterator::upper_bound, after);
+    }
+    seastar::future<> lower_bound(const std::string &to) final {
+      return proxy(&OmapIterator::lower_bound, to);
+    }
+    bool valid() const final {
+      return impl->valid();
+    }
+    seastar::future<> next() final {
+      return proxy(&OmapIterator::next);
+    }
+    std::string key() final {
+      return impl->key();
+    }
+    ceph::buffer::list value() final {
+      return impl->value();
+    }
+    int status() const final {
+      return impl->status();
+    }
+    ~_OmapIterator() = default;
+  };
+
+
+public:
+  ShardedStoreProxy(T *t)
+    : core(seastar::this_shard_id()),
+      impl(t) {}
+  template <typename... Args>
+  ShardedStoreProxy(Args&&... args)
+    : core(seastar::this_shard_id()),
+      impl(std::make_unique<T>(std::forward<Args>(args)...)) {}
+  ~ShardedStoreProxy() = default;
+
+  // no copying
+  explicit ShardedStoreProxy(const ShardedStoreProxy &o) = delete;
+  const ShardedStoreProxy &operator=(const ShardedStoreProxy &o) = delete;
+
+  seastar::future<> start() final { return proxy(&T::start); }
+  seastar::future<> stop() final { return proxy(&T::stop); }
+  mount_ertr::future<> mount() final {
+    auto ret = seastar::smp::submit_to(
+      core,
+      [this] {
+	auto ret = impl->mount(
+	).safe_then([this] {
+	  fsid = impl->get_fsid();
+	  max_attr = impl->get_max_attr_name_length();
+	  return seastar::now();
+	});
+	return std::move(ret).to_base();
+      });
+    return mount_ertr::future<>(std::move(ret));
+  }
+  seastar::future<> umount() final { return proxy(&T::umount); }
+  mkfs_ertr::future<> mkfs(uuid_d new_osd_fsid) final {
+    return proxy(&T::mkfs, new_osd_fsid);
+  }
+  seastar::future<store_statfs_t> stat() const final {
+    return crimson::submit_to(core, [this] { return impl->stat(); });
+  }
+  read_errorator::future<ceph::bufferlist> read(
```

Comment:
From seastar/include/core/memory.hh

```
/// The seastar memory allocator splits system memory into a pool per
/// logical core (lcore).  Memory allocated one an lcore should be freed
/// on the same lcore; failing to do so carries a severe performance
/// penalty.  It is possible to share memory with another core, but this
/// should be limited to avoid cache coherency traffic.
/// You can obtain the memory layout of the current shard with
/// \ref get_memory_layout().
```

Freeing memory on another core appears to be safe, just not performant.  We can improve this later, though I suspect that for buffers that actually need to move between cores, some overhead is necessary.

Reply:
Yeah, seastar is able to handle malloc/free from inconsistent cores: https://github.com/scylladb/seastar/blob/cb8d2b14c322cbd42aecacb3ec686f5d82393d93/src/core/memory.cc#L961-L978

So it is correct to move unique_ptrs and copy thread-safe shared_ptrs across cores.

Only concerning correctness, our code should not share ptrs with thread-unsafe ref-counters across cores, currently I see:
* bufferlist containing non-foreign seastar::temporary_buffer, probably we should by-default to create buffers using `buffer::create_foreign(temporary_buffer&&)`?
* CollectionRef is thread-unsafe.

Reply:
Yup, didn't notice CollectionRef particularly.  Will fix tomorrow.

Reply:
Eh, for now I'll just switch it to a thread_safe_counter.  I'll follow up with a PR replacing CollectionRef in the interface with a coll_t and eliminate this as a problem https://tracker.ceph.com/issues/57653 .

Reply:
> Only concerning correctness, our code should not share ptrs with thread-unsafe ref-counters across cores, currently I see:
>
>  *  bufferlist containing non-foreign seastar::temporary_buffer, probably we should by-default to create buffers using `buffer::create_foreign(temporary_buffer&&)`?

I think the `bufferlist` ref-counts atomically already for the sake of `AlienStore`. Otherwise we would need a `bptr` translation there.

Reply:
Since the bufferptr owns the only reference to the buffer, I don't think it's a problem.  bufferptr's themselves are also safe because ceph::buffer::raw uses, unconditionally, a std::atomic for nref.

Reply:
`temporary_buffer` is internally shareable with a thread-unsafe ref-counter: https://github.com/scylladb/seastar/blob/master/include/seastar/core/deleter.hh#L111

I'm not sure whether seastar internally shares the buffer when pass `temporary_buffer` to Crimson. If so, releasing ``temporary_buffer` in another core might be an issue.

Reply:
True, I suppose we can't ensure that seastar won't retain a reference to the temporary_buffer.  For now, I think the right answer is to rename create -> create_local and create_foreign -> create.  Users that need the extra efficiency can swap to create_local as needed.

Reply:
https://github.com/ceph/ceph/pull/48217